src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -465,6 +426,60 @@ private bool TryRegister(SafeSocketHandle socket, IntPtr handle, out Interop.Err
             return error == Interop.Error.SUCCESS;
         }
 
+        // An on-stack cache of SocketAsyncEngine members needed to process epoll events.
+        // The event handling logic is delegated to a non-inlined processing method
+        // to make sure the runtime doesn't retain internal copies of SocketAsyncContext longer than necessary,
+        // so unreferenced Socket instances always get collected by the GC.
+        // See discussion: https://github.com/dotnet/runtime/issues/37064
+        private struct SocketEventHandler
+        {
+            public bool Shutdown { get; private set; }
+            public Interop.Sys.SocketEvent* Buffer { get; }
+
+            private ConcurrentDictionary<IntPtr, SocketAsyncContextWrapper> _handleToContextMap;
+            private ConcurrentQueue<SocketIOEvent> _eventQueue;
+            private IntPtr _shutdownHandle;
+
+            public SocketEventHandler(SocketAsyncEngine engine)
+            {
+                Shutdown = false;
+                Buffer = engine._buffer;
+                _handleToContextMap = engine._handleToContextMap;
+                _eventQueue = engine._eventQueue;
+                _shutdownHandle = ShutdownHandle;
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            public bool HandleSocketEvents(int numEvents)
+            {
+                SocketAsyncContext? context = null;
+                bool enqueuedEvent = false;
+                foreach (var socketEvent in new ReadOnlySpan<Interop.Sys.SocketEvent>(Buffer, numEvents))
+                {
+                    IntPtr handle = socketEvent.Data;
+
+                    if (_handleToContextMap.TryGetValue(handle, out SocketAsyncContextWrapper contextWrapper) && (context = contextWrapper.Context) != null)
```

Comment:
I don't think this `!= null` check is needed. When `TryGetValue` returns `true`, there should be a `context`, right?

`context` variable may be assigned under the `if` block:
```
SocketAsyncContext context = contextWrapper.Context!;
```

Reply:
Seems logical. @adamsitnik I think you authored this in #35800. Is there anything we are missing?

Reply:
@antonfirsov @tmds I just wanted to be super defensive as the code was previously:

```cs
handleToContextMap.TryGetValue(handle, out SocketAsyncContext? context);
if (context != null)
```

but looking at it again I have no idea why it was not just checking if `TryGetValue` returned true. We should most probably simplify it and remove the extra check